Sequential Tree Sampler (STS)
=============================

The sequential tree sampler implements a prototype of online phylogenetics, updating a posterior distribution generated by MrBayes with new sequences.
The algorithm has been described and its performance evaluated in [a manuscript][preprint].
The scripts used to generate the figures can be found [here](https://github.com/OnlinePhylo/sts-online-experiments).

## Dependencies

* [smctc][smctc] - included as git submodule (`git submodule update --init`)
* [lcfit][lcfit] - included as git submodule (`git submodule update --init`)
* [Bio++ version 2.2.0][bpp] `core`, `seq`, and `phyl` modules. Note that debian & ubuntu up to 16.04 include v2.1.0 which is too old. Bio++ should be installed from source using the `bpp-setup.sh` script on these systems. Alternatively, the source code of version 2.3.0 for each module can be dowloaded from [github](https://github.com/BioPP) in the _releases_ section
* [cmake][cmake]
* [gsl version 1.16][gsl] Note that gsl v2 is not currently supported.
* [nlopt][nlopt]
* [boost] [boost]
* [beagle][beagle] version 2.1 (Optional)
* [google test][gtest] this is libgtest on debian/ubuntu (Optional)

## Compiling

1. Install dependencies
2. run `make`

Binaries will be build in `_build/release`

## Adding taxa to an existing posterior

The tool `sts-online` adds taxa to an existing posterior tree sample.
`sts-online` operates on a fasta file and tree file in nexus format.
The fasta file must contain an alignment with a superset of the taxa in the tree file.

Currently only the Jukes-Cantor model is supported.

### Example invocation

    sts-online -b 250 -p 2 --proposal-method lcfit 10taxon-01.fasta 10tax_trim_t1.t 10tax_trim_t1.sts.json

In this example, we use an alignment containing 10 sequences and a posterior sample of trees generated by MrBayes with an alignment that does not contain the sequence labeled `t1`.
`sts-online` ignores the first 250 trees from `50tax_trim.run1.t` and  uses a particle factor of 2. The `10tax_trim_t1.sts.json` file will contain the updated trees.



[preprint]: http://biorxiv.org/content/early/2017/06/02/145219
[smctc]: http://www2.warwick.ac.uk/fac/sci/statistics/staff/academic-research/johansen/smctc/
[lcfit]: http://github.com/matsengrp/lcfit/
[beagle]: https://code.google.com/p/beagle-lib/
[bpp]: http://biopp.univ-montp2.fr/
[cmake]: http://www.cmake.org/
[gsl]: https://www.gnu.org/software/gsl/
[nlopt]: http://ab-initio.mit.edu/wiki/index.php/NLopt
[boost]: https://www.boost.org
[gtest]: https://github.com/google/googletest

